# 📊 Desafio 07 - Comparação Percentual por Curso | #7DaysOfCode - Alura

Neste último desafio da jornada, nosso objetivo é apoiar a **diretoria da biblioteca** na criação de um plano de ação focado nos **alunos da pós-graduação**, com base nos dados históricos de empréstimos dos cursos.

---

### 🎯 O desafio

A diretoria deseja saber:
> **"Como evoluíram os empréstimos de cada curso de pós-graduação entre 2017, 2018, 2019 e 2022?"**

A partir disso, será possível identificar cursos com tendências de crescimento ou queda, e investir em ações de incentivo mais personalizadas.

---

### 📌 A tarefa

Precisamos:

1. Filtrar os alunos da **pós-graduação**.
2. Realizar o **merge** com os empréstimos para identificar quais registros pertencem a esses alunos.
3. Considerar apenas os anos **2017, 2018, 2019**.
4. Utilizar uma **previsão de 2022** já fornecida para compor a análise.
5. Criar uma **tabela comparativa** por curso, mostrando as diferenças percentuais:
   - de **2017 para 2018**
   - de **2018 para 2019**
   - de **2019 para 2022**
6. Gerar a tabela final em **HTML estilizado** com:
   - ✅ Cores distintas para positivos (verde) e negativos (vermelho)
   - ✅ Cabeçalho visualmente destacado
   - ✅ Cursos formatados corretamente
   - ✅ Remoção do índice

---

### 🧪 Vamos começar com:

- 📥 Importação dos dados tratados (empréstimos e alunos)
- 🧼 Filtro dos **alunos da pós-graduação**
- 🔗 Merge com os empréstimos

## 📥 Importação dos Dados de Usuários

Vamos importar os dados obtidos até aqui:

- Um arquivo `.csv` com os dados de empréstimos da biblioteca
- Um arquivo `.cvs` com os dados dos alunos

Esses dados serão combinados em um único DataFrame para posterior análise que será entregue à diretoria.

In [4]:
# Realizando Importações

# 📚 Importando bibliotecas essenciais
import pandas as pd
import numpy as np

# Importando os dados de Empréstimos
df_emprestimos = pd.read_csv('../desafio04/dataset_completo.csv')

# Importando os dados dos Alunos
df_alunos = pd.read_csv('../desafio06/dados/df_alunos.csv')

display(df_emprestimos.head(3), df_emprestimos.value_counts(), df_alunos.head(3), df_alunos.value_counts())

C:\Users\Jhonny\AppData\Local\Temp\ipykernel_19116\1408394137.py:8: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  df_emprestimos = pd.read_csv('../desafio04/dataset_completo.csv')


id_emprestimo codigo_barras           data_renovacao  \
0         709684       L095049                      NaN   
1         709685       L167050                      NaN   
2         709686    2006017618  2010-01-26 08:07:01.738   

           data_emprestimo           data_devolucao  matricula_ou_siape  \
0  2010-01-04 07:44:10.721  2010-01-05 16:26:12.662        2.008023e+09   
1  2010-01-04 07:44:10.750  2010-01-12 07:34:13.934        2.008023e+09   
2  2010-01-04 08:08:44.081  2010-02-25 07:36:25.800        2.008112e+09   

     tipo_vinculo_usuario  id_exemplar            colecao  \
0      ALUNO DE GRADUAÇÃO      13259.0  Acervo Circulante   
1      ALUNO DE GRADUAÇÃO      70865.0  Acervo Circulante   
2  ALUNO DE PÓS-GRADUAÇÃO     195347.0  Acervo Circulante   

                                          biblioteca status_material  \
0                     Biblioteca Central Zila Mamede         REGULAR   
1                     Biblioteca Central Zila Mamede         REGULAR   
2  Biblioteca Setorial Prof. Rodolfo Helinski - E...         REGULAR   

   localizacao          classe_cdu  
0          630  Ciências aplicadas  
1          647  Ciências aplicadas  
2          640  Ciências aplicadas

id_emprestimo  codigo_barras  data_renovacao           data_emprestimo          data_devolucao              matricula_ou_siape  tipo_vinculo_usuario    id_exemplar  colecao            biblioteca                                                                                status_material  localizacao  classe_cdu        
709686         2006017618     2010-01-26 08:07:01.738  2010-01-04 08:08:44.081  2010-02-25 07:36:25.800     2.008112e+09        ALUNO DE PÓS-GRADUAÇÃO  195347.0     Acervo Circulante  Biblioteca Setorial Prof. Rodolfo Helinski - Escola Agrícola de Jundiaí - EAJ  - Macaiba  REGULAR          640          Ciências aplicadas    1
784698         2009052756     2010-06-14 20:33:19.993  2010-05-31 15:58:17.377  2010-06-23 15:19:04.000     2.007293e+08        ALUNO DE GRADUAÇÃO      282148.0     Acervo Circulante  Biblioteca Central Zila Mamede                                                            REGULAR          605          Ciências aplicadas    1
784685         L11402

Unnamed: 0  matricula_ou_siape tipo_vinculo_usuario                curso
0           0         190001688.0   ALUNO DE GRADUAÇÃO           MATEMÁTICA
1           1         199014489.0   ALUNO DE GRADUAÇÃO  CIÊNCIAS BIOLÓGICAS
2           2         199735674.0   ALUNO DE GRADUAÇÃO           MATEMÁTICA

Unnamed: 0  matricula_ou_siape  tipo_vinculo_usuario    curso             
0           1.900017e+08        ALUNO DE GRADUAÇÃO      MATEMÁTICA            1
59976       2.016016e+10        ALUNO DE GRADUAÇÃO      COMUNICAÇÃO SOCIAL    1
59984       2.018004e+10        ALUNO DE GRADUAÇÃO      AGRONOMIA             1
59983       2.017012e+10        ALUNO DE GRADUAÇÃO      ENFERMAGEM            1
59982       2.016010e+10        ALUNO DE GRADUAÇÃO      ENGENHARIA CIVIL      1
                                                                             ..
29987       2.013029e+09        ALUNO DE GRADUAÇÃO      FILOSOFIA             1
29986       2.012928e+09        ALUNO DE GRADUAÇÃO      MATEMÁTICA            1
29985       2.013040e+09        ALUNO DE GRADUAÇÃO      BIBLIOTECONOMIA       1
29984       2.013034e+09        ALUNO DE GRADUAÇÃO      FILOSOFIA             1
89967       2.020101e+10        ALUNO DE PÓS-GRADUAÇÃO  DESIGN                1
Name: count, Length: 89968, dtype: int64

## 🧼 Pré-processamento dos Dados

Antes de realizarmos qualquer análise, precisamos garantir que nossos dados estejam prontos para uso. 

Nesta etapa, faremos:
- Conversão de colunas de **data** para o tipo `datetime`, essencial para filtros por ano;
- Padronização da coluna `matricula_ou_siape` para o tipo **string**, já que será usada para o `merge` entre os dados de alunos e de empréstimos;
- Verificação inicial com `display()` para visualizar a estrutura dos dados carregados.

Essas etapas são cruciais para evitar erros futuros ao manipular os dados, aplicar filtros por ano e agrupar por curso.

In [6]:
# Primeiras Conversões

# Garante que a coluna está no tipo datetime
df_emprestimos['data_emprestimo'] = pd.to_datetime(df_emprestimos['data_emprestimo'], errors='coerce')

# Transformando a matrícula em string
df_emprestimos['matricula_ou_siape'] = df_emprestimos['matricula_ou_siape'].astype(str)
df_alunos['matricula_ou_siape'] = df_alunos['matricula_ou_siape'].astype(str)

display(df_emprestimos.head(3), df_emprestimos.value_counts(), df_alunos.head(3), df_alunos.value_counts())

id_emprestimo codigo_barras           data_renovacao  \
0         709684       L095049                      NaN   
1         709685       L167050                      NaN   
2         709686    2006017618  2010-01-26 08:07:01.738   

          data_emprestimo           data_devolucao matricula_ou_siape  \
0 2010-01-04 07:44:10.721  2010-01-05 16:26:12.662       2008023265.0   
1 2010-01-04 07:44:10.750  2010-01-12 07:34:13.934       2008023265.0   
2 2010-01-04 08:08:44.081  2010-02-25 07:36:25.800       2008112114.0   

     tipo_vinculo_usuario  id_exemplar            colecao  \
0      ALUNO DE GRADUAÇÃO      13259.0  Acervo Circulante   
1      ALUNO DE GRADUAÇÃO      70865.0  Acervo Circulante   
2  ALUNO DE PÓS-GRADUAÇÃO     195347.0  Acervo Circulante   

                                          biblioteca status_material  \
0                     Biblioteca Central Zila Mamede         REGULAR   
1                     Biblioteca Central Zila Mamede         REGULAR   
2  Biblioteca Setorial Prof. Rodolfo Helinski - E...         REGULAR   

   localizacao          classe_cdu  
0          630  Ciências aplicadas  
1          647  Ciências aplicadas  
2          640  Ciências aplicadas

id_emprestimo  codigo_barras  data_renovacao           data_emprestimo          data_devolucao              matricula_ou_siape  tipo_vinculo_usuario    id_exemplar  colecao            biblioteca                                                                                status_material  localizacao  classe_cdu        
709686         2006017618     2010-01-26 08:07:01.738  2010-01-04 08:08:44.081  2010-02-25 07:36:25.800     2008112114.0        ALUNO DE PÓS-GRADUAÇÃO  195347.0     Acervo Circulante  Biblioteca Setorial Prof. Rodolfo Helinski - Escola Agrícola de Jundiaí - EAJ  - Macaiba  REGULAR          640          Ciências aplicadas    1
784678         2009014131     2010-06-15 09:04:45.962  2010-05-31 15:42:26.842  2010-06-30 14:26:59.000     200715089.0         ALUNO DE GRADUAÇÃO      267092.0     Acervo Circulante  Biblioteca Central Zila Mamede                                                            REGULAR          683          Ciências aplicadas    1
784649         200903

Unnamed: 0 matricula_ou_siape tipo_vinculo_usuario                curso
0           0        190001688.0   ALUNO DE GRADUAÇÃO           MATEMÁTICA
1           1        199014489.0   ALUNO DE GRADUAÇÃO  CIÊNCIAS BIOLÓGICAS
2           2        199735674.0   ALUNO DE GRADUAÇÃO           MATEMÁTICA

Unnamed: 0  matricula_ou_siape  tipo_vinculo_usuario    curso             
0           190001688.0         ALUNO DE GRADUAÇÃO      MATEMÁTICA            1
59976       20160157168.0       ALUNO DE GRADUAÇÃO      COMUNICAÇÃO SOCIAL    1
59984       20180037452.0       ALUNO DE GRADUAÇÃO      AGRONOMIA             1
59983       20170117620.0       ALUNO DE GRADUAÇÃO      ENFERMAGEM            1
59982       20160104763.0       ALUNO DE GRADUAÇÃO      ENGENHARIA CIVIL      1
                                                                             ..
29987       2013029270.0        ALUNO DE GRADUAÇÃO      FILOSOFIA             1
29986       2012928188.0        ALUNO DE GRADUAÇÃO      MATEMÁTICA            1
29985       2013040348.0        ALUNO DE GRADUAÇÃO      BIBLIOTECONOMIA       1
29984       2013033737.0        ALUNO DE GRADUAÇÃO      FILOSOFIA             1
89967       20201014993.0       ALUNO DE PÓS-GRADUAÇÃO  DESIGN                1
Name: count, Length: 89968, dtype: int64

In [7]:
# Fazendo o filtro do `df_alunos` pelo tipo de vínculo do aluno
df_pos = df_alunos[
    (df_alunos['tipo_vinculo_usuario'] == 'ALUNO DE PÓS-GRADUAÇÃO')] # Toma somente os alunos da graduação

## 🔗 Realizando o Merge entre Empréstimos e Alunos de Pós-graduação

Agora que já temos nossos dados de empréstimos padronizados e os alunos de **pós-graduação** devidamente filtrados, o próximo passo é unir essas duas bases de dados.

Usaremos a coluna `matricula_ou_siape` como chave de ligação entre os dois DataFrames:

- **df_pos**: contém os dados dos alunos da pós-graduação.
- **df_emprestimos**: contém todos os empréstimos registrados, de alunos de todos os níveis.

### O objetivo do `merge`:
Criar um novo DataFrame contendo apenas os **empréstimos realizados por alunos da pós-graduação**.

O tipo de junção escolhido será `inner`, para manter apenas as matrículas que estão **presentes nas duas bases** (ou seja, que realizaram empréstimos).

In [ ]:
# Mesclando os DataFrames
df_emprestimos_pos = pd.merge(
    df_emprestimos,  # dados de empréstimos
    df_alunos,  # dados de alunos da pós-graduação
    on='matricula_ou_siape', # vínculo entre os DataFrames
    how='inner'  # inner porque queremos apenas as matrículas que têm correspondência
)
df_emprestimos_pos.head(5)

id_emprestimo codigo_barras           data_renovacao  \
0         709684       L095049                      NaN   
1         709685       L167050                      NaN   
2         709686    2006017618  2010-01-26 08:07:01.738   
3         709687       L184117  2010-01-18 11:07:46.470   
4         709688    2007027500  2010-01-19 15:59:26.464   

          data_emprestimo           data_devolucao matricula_ou_siape  \
0 2010-01-04 07:44:10.721  2010-01-05 16:26:12.662       2008023265.0   
1 2010-01-04 07:44:10.750  2010-01-12 07:34:13.934       2008023265.0   
2 2010-01-04 08:08:44.081  2010-02-25 07:36:25.800       2008112114.0   
3 2010-01-04 08:24:21.284  2010-02-03 08:58:45.692        200721119.0   
4 2010-01-04 08:27:33.569  2010-02-03 13:06:30.662       2009045971.0   

   tipo_vinculo_usuario_x  id_exemplar            colecao  \
0      ALUNO DE GRADUAÇÃO      13259.0  Acervo Circulante   
1      ALUNO DE GRADUAÇÃO      70865.0  Acervo Circulante   
2  ALUNO DE PÓS-GRADUAÇÃO     195347.0  Acervo Circulante   
3      ALUNO DE GRADUAÇÃO     131639.0  Acervo Circulante   
4      ALUNO DE GRADUAÇÃO     225646.0  Acervo Circulante   

                                          biblioteca status_material  \
0                     Biblioteca Central Zila Mamede         REGULAR   
1                     Biblioteca Central Zila Mamede         REGULAR   
2  Biblioteca Setorial Prof. Rodolfo Helinski - E...         REGULAR   
3                     Biblioteca Central Zila Mamede         REGULAR   
4                     Biblioteca Central Zila Mamede         REGULAR   

   localizacao          classe_cdu  Unnamed: 0  tipo_vinculo_usuario_y  \
0          630  Ciências aplicadas        4290      ALUNO DE GRADUAÇÃO   
1          647  Ciências aplicadas        4290      ALUNO DE GRADUAÇÃO   
2          640  Ciências aplicadas       70528  ALUNO DE PÓS-GRADUAÇÃO   
3          609  Ciências aplicadas        2525      ALUNO DE GRADUAÇÃO   
4          686  Ciências aplicadas        6856      ALUNO DE GRADUAÇÃO   

                                   curso  
0                     ENGENHARIA QUÍMICA  
1                     ENGENHARIA QUÍMICA  
2                 CIÊNCIAS ODONTOLÓGICAS  
3                              GEOGRAFIA  
4  ANÁLISE E DESENVOLVIMENTO DE SISTEMAS

## 📆 Filtrando o Período de Análise (2017 a 2019)

Com o `DataFrame` já contendo somente os empréstimos realizados por alunos da pós-graduação, agora vamos restringir a análise apenas aos anos de **2017, 2018 e 2019**.

Esse recorte é importante porque:
- **2020 e 2021** foram anos atípicos por conta da pandemia e não devem ser considerados;
- A diretoria está interessada na **comparação histórica** até 2019;
- Posteriormente, incluiremos os dados de previsão de **2022**.

Vamos então aplicar esse filtro com base na coluna `data_emprestimo`, que já está no formato `datetime`.

In [9]:
# Criando a coluna com o ano para facilitar o filtro
df_emprestimos_pos['ano'] = df_emprestimos_pos['data_emprestimo'].dt.year

# Filtrando apenas os anos desejados
df_emprestimos_pos = df_emprestimos_pos[
    df_emprestimos_pos['ano'].isin([2017, 2018, 2019])
]

df_emprestimos_pos.head(5)

id_emprestimo codigo_barras           data_renovacao  \
1415739        9405484    2011010806  2017-01-13 07:22:40.736   
1415740        9405488       L078593  2017-01-31 07:31:36.960   
1415741        9405492    2010040100  2017-01-31 07:31:36.902   
1415742        9405496    2013105705  2017-01-31 07:31:36.843   
1415743        9405500       L057436  2017-01-31 07:30:36.293   

                data_emprestimo data_devolucao matricula_ou_siape  \
1415739 2017-01-02 08:01:14.113            NaN       2014035816.0   
1415740 2017-01-02 08:29:46.657            NaN      20161011168.0   
1415741 2017-01-02 08:29:46.722            NaN      20161011168.0   
1415742 2017-01-02 08:29:46.794            NaN      20161011168.0   
1415743 2017-01-02 08:38:40.009            NaN      20161010760.0   

         tipo_vinculo_usuario_x  id_exemplar            colecao  \
1415739      ALUNO DE GRADUAÇÃO     915480.0  Acervo Circulante   
1415740  ALUNO DE PÓS-GRADUAÇÃO          NaN                NaN   
1415741  ALUNO DE PÓS-GRADUAÇÃO    1051275.0  Acervo Circulante   
1415742  ALUNO DE PÓS-GRADUAÇÃO    1097159.0  Acervo Circulante   
1415743  ALUNO DE PÓS-GRADUAÇÃO          NaN                NaN   

                                                biblioteca status_material  \
1415739  Biblioteca Setorial Profª. Maria José Mamede G...         REGULAR   
1415740                                                NaN             NaN   
1415741                     Biblioteca Central Zila Mamede         REGULAR   
1415742                     Biblioteca Central Zila Mamede         REGULAR   
1415743                                                NaN             NaN   

         localizacao                             classe_cdu  Unnamed: 0  \
1415739          689                     Ciências aplicadas       36657   
1415740            0  Generalidades. Ciência e conhecimento       81525   
1415741          651                     Ciências aplicadas       81525   
1415742          380                       Ciências sociais       81525   
1415743            0  Generalidades. Ciência e conhecimento       81526   

         tipo_vinculo_usuario_y             curso   ano  
1415739      ALUNO DE GRADUAÇÃO  ENGENHARIA CIVIL  2017  
1415740  ALUNO DE PÓS-GRADUAÇÃO         FILOSOFIA  2017  
1415741  ALUNO DE PÓS-GRADUAÇÃO         FILOSOFIA  2017  
1415742  ALUNO DE PÓS-GRADUAÇÃO         FILOSOFIA  2017  
1415743  ALUNO DE PÓS-GRADUAÇÃO    BIOINFORMÁTICA  2017

## 📊 Reorganizando os dados: formato longo para formato largo

Até este ponto, os dados estavam organizados no formato "longo", ou seja, cada linha representava um empréstimo com o respectivo ano indicado na coluna `ano`.

Para podermos calcular a variação percentual ano a ano — como solicitado pela diretoria — precisamos converter esses dados para o formato "largo", com **uma coluna para cada ano** (2017, 2018, 2019), e os cursos como índice.

Utilizamos a função `pivot_table()` do Pandas para transformar a base nesse novo formato, contando a quantidade de empréstimos por curso e por ano. Após isso, faremos um `merge` com a tabela `df_2022`, que contém a previsão de empréstimos para o ano de 2022.

In [26]:
# Criando uma tabela com os cursos como índice e anos como colunas
tabela_anos = df_emprestimos_pos.pivot_table(
    index='curso',
    columns='ano',
    values='id_emprestimo',
    aggfunc='count',
    fill_value=0
).reset_index()

tabela_anos.head(5)

ano                                  curso   2017  2018  2019
0                            ADMINISTRAÇÃO  10419  8915  7614
1                                AGRONOMIA   7834  7120  5880
2    ANÁLISE E DESENVOLVIMENTO DE SISTEMAS     16    11     0
3                  ARQUITETURA E URBANISMO  10746  9183  8114
4                            ARTES CÊNICAS   2217  2162  1895

In [33]:
tabela_anos[2017].dtype

dtype('int64')

## 🌐 Importando a Tabela de Previsão (2022)

Como parte da análise solicitada pela diretoria, iremos incorporar os dados de **previsão para o ano de 2022**, gerados por outra equipe.

A tabela está disponível em formato `.html`, hospedada em um repositório no GitHub. Utilizaremos as bibliotecas `requests` e `pandas.read_html()` para fazer a leitura e extração da tabela.

A tabela será armazenada localmente no diretório `dados/`, mantendo o mesmo padrão de organização usado nos desafios anteriores.

In [15]:
!pip install -q lxml
!pip install -q html5lib

In [16]:
from pathlib import Path
import requests
import pandas as pd

# Criando diretório local para armazenar os dados
pasta_dados = Path('dados')
pasta_dados.mkdir(parents=True, exist_ok=True)

# URL do arquivo HTML (link raw)
url_html = "https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_7-Apresentando_resultados_em_HTML/Dataset/previsao"

# Realizando o download
r_html = requests.get(url_html)
with open("dados/dados_2022.html", "wb") as f:
    f.write(r_html.content)

# Lendo a tabela HTML com Pandas
df_2022 = pd.read_html('dados/dados_2022.html')[0]

# Visualizando os dados
df_2022.head()

ValueError: No tables found

## 📄 Alternativa: Inserção manual dos dados de 2022

Durante a tentativa de leitura da tabela `.html`, constatamos que o arquivo não contém marcações HTML válidas com `<table>`, impossibilitando o uso direto do `pandas.read_html()`.

Como alternativa, replicaremos os dados manualmente com base na previsão fornecida no link:

📎 [Tabela de Previsão - 2022 (GitHub)](https://raw.githubusercontent.com/FranciscoFoz/7_Days_of_Code_Alura-Python-Pandas/main/Dia_7-Apresentando_resultados_em_HTML/Dataset/previsao)

O DataFrame criado contém as projeções de empréstimos por curso da pós-graduação para o ano de 2022.

In [19]:
# Criando manualmente a tabela com base na previsão

dados_2022 = {
    'curso': [
        'ADMINISTRAÇÃO',
        'ARQUITETURA E URBANISMO',
        'ARTES CÊNICAS',
        'BIOINFORMÁTICA',
        'BIOQUÍMICA',
        'CIÊNCIA E ENGENHARIA DE MATERIAIS',
        'CIÊNCIA TECNOLOGIA E INOVAÇÃO',
        'CIÊNCIAS DA SAÚDE',
        'CIÊNCIAS ODONTOLÓGICAS',
        'DESIGN',
        'ENGENHARIA CIVIL',
        'ENGENHARIA DE SOFTWARE',
        'FILOSOFIA',
        'GESTÃO DA INFORMAÇÃO E DO CONHECIMENTO',
    ],
    'previsao_2022': [1905, 1896, 1920, 2302, 2090, 2010, 1960, 1730, 2405, 1900, 1920, 1890, 1903, 1855]
}

df_2022 = pd.DataFrame(dados_2022)
df_2022.head()


curso  previsao_2022
0            ADMINISTRAÇÃO           1905
1  ARQUITETURA E URBANISMO           1896
2            ARTES CÊNICAS           1920
3           BIOINFORMÁTICA           2302
4               BIOQUÍMICA           2090

In [31]:
df_2022['previsao_2022'].dtype

dtype('int64')

## 🔄 Integração com dados de previsão 2022

Com os dados históricos de empréstimos entre 2017 e 2019 já agrupados por curso, e os dados de previsão de 2022 inseridos manualmente, o próximo passo é integrá-los.

Será feita uma junção (`merge`) entre os dois DataFrames, utilizando a coluna `curso` como chave, garantindo consistência nos nomes e formatos. Em seguida, as colunas serão organizadas para possibilitar o cálculo da **diferença percentual ano a ano**.

As diferenças percentuais serão calculadas entre os seguintes pares de anos:
- 2017 → 2018
- 2018 → 2019
- 2019 → 2022 (previsão)

Essas análises fornecerão insumos estratégicos para a diretoria da biblioteca no desenvolvimento de ações direcionadas ao público da pós-graduação.

In [27]:
# Mesclando os DataFrames
df_completo = pd.merge(
    tabela_anos,       # DataFrame com colunas 2017, 2018, 2019
    df_2022,         # DataFrame com a previsão
    on='curso',      # Chave comum
    how='inner'      # Apenas os cursos com correspondência nos dois
)
df_completo.head(5)

curso   2017  2018  2019  previsao_2022
0            ADMINISTRAÇÃO  10419  8915  7614           1905
1  ARQUITETURA E URBANISMO  10746  9183  8114           1896
2            ARTES CÊNICAS   2217  2162  1895           1920
3           BIOINFORMÁTICA   2308  1990  1889           2302
4               BIOQUÍMICA   2389  2021  2126           2090

## 📊 Cálculo das Diferenças Percentuais (2017–2022)

Com o DataFrame completo em mãos, contendo os valores de empréstimos realizados entre 2017 e 2019 e a projeção para 2022, o próximo passo é calcular a variação percentual entre os anos.

Essas diferenças revelam padrões de crescimento ou declínio nos empréstimos realizados pelos alunos da pós-graduação, curso a curso. Serão geradas três colunas:

- `2017–2018`
- `2018–2019`
- `2019–2022`

Cada valor percentual será analisado para posterior estilização em HTML, destacando aumentos em **verde** e quedas em **vermelho**, conforme requisitado pela equipe de Front-end.

In [34]:
df_completo['2017-2018'] = ((df_completo[2018] - df_completo[2017]) / df_completo[2017]) * 100
df_completo['2018-2019'] = ((df_completo[2019] - df_completo[2018]) / df_completo[2018]) * 100
df_completo['2019-2022'] = ((df_completo['previsao_2022'] - df_completo[2019]) / df_completo[2019]) * 100

df_completo.head(5)

curso   2017  2018  2019  previsao_2022  2017-2018  \
0            ADMINISTRAÇÃO  10419  8915  7614           1905 -14.435167   
1  ARQUITETURA E URBANISMO  10746  9183  8114           1896 -14.544947   
2            ARTES CÊNICAS   2217  2162  1895           1920  -2.480830   
3           BIOINFORMÁTICA   2308  1990  1889           2302 -13.778163   
4               BIOQUÍMICA   2389  2021  2126           2090 -15.403935   

   2018-2019  2019-2022  
0 -14.593382 -74.980299  
1 -11.641076 -76.632980  
2 -12.349676   1.319261  
3  -5.075377  21.863420  
4   5.195448  -1.693321

## 💻 Geração da Tabela HTML Estilizada

Com as variações percentuais de empréstimos entre os anos calculadas, criamos agora uma tabela final com os cursos e suas respectivas diferenças entre:

- 2017 e 2018
- 2018 e 2019
- 2019 e 2022 (previsão)

Esta tabela será formatada para visualização web, com destaque em cores para facilitar a interpretação:

- **Verde** para crescimento;
- **Vermelho** para queda;
- Percentuais com **duas casas decimais e símbolo “%”**;
- Cursos com **formatação capitalizada** (ex: “Filosofia”).

In [41]:
# Etapa 1: Selecionando apenas as colunas desejadas
tabela_final = df_completo[['curso', '2017-2018', '2018-2019', '2019-2022']].copy()

# Padronizando o nome dos cursos
tabela_final['curso'] = tabela_final['curso'].str.title()

# Formatando os valores numéricos com %. Dica: use apply + format
for coluna in ['2017-2018', '2018-2019', '2019-2022']:
    tabela_final[coluna] = tabela_final[coluna].apply(lambda x: f'{x:.2f}%')

## 🎨 Estilizando a Tabela Final

Para apoiar a equipe de Front-End na publicação dos resultados em uma página web, a diretoria solicitou que a tabela com as variações percentuais de empréstimos seja exportada em HTML, com estilo visual padronizado.

As exigências da equipe foram:

- Não exibir a numeração de índice;
- Os nomes dos cursos devem estar capitalizados apenas na primeira letra;
- Os percentuais devem conter o símbolo `%` e duas casas decimais;
- Os valores positivos devem aparecer em **verde** e os negativos em **vermelho**;
- O arquivo deve conter estilos CSS embutidos, conforme padrão visual já definido.

A seguir, aplicamos esses requisitos utilizando os recursos de estilização do `pandas` com `.style` e salvamos o resultado em HTML.

In [44]:
from IPython.display import HTML, display

# Função para colorir as células com base no valor
def colorir_valores(val):
    try:
        valor = float(val.replace('%', '').replace(',', '.'))
        if valor > 0:
            return 'color: green; font-weight: bold;'
        elif valor < 0:
            return 'color: red; font-weight: bold;'
        else:
            return ''
    except:
        return ''

# Estilo do cabeçalho e corpo da tabela
estilo_css = """
<style>
    table {
        font-family: Arial, sans-serif;
        border-collapse: collapse;
        margin: 25px 0;
        font-size: 1rem;
        width: 100%;
        box-shadow: 0 0 1rem gray;
    }
    thead tr {
        background-color: #001692;
        color: whitesmoke;
        text-align: center;
        font-size: 1.4rem;
        font-weight: bold;
        border-radius: 0.25rem;
    }
    tbody td {
        padding: 0.5rem;
        border-bottom: 0.1rem solid lightgray;
        text-align: left;
        font-weight: bold;
        font-size: 1rem;
    }
</style>
"""

# Aplicando estilo e colorização com pandas Styler
styled_tabela = (
    tabela_final
    .style
    .applymap(colorir_valores, subset=['2017-2018', '2018-2019', '2019-2022'])
    .hide(axis='index')  # remove o índice
)

# Exportando a tabela estilizada para HTML
with open("dados/tabela_variacoes.html", "w", encoding="utf-8") as f:
    f.write(estilo_css + styled_tabela.to_html())

display(HTML(estilo_css + styled_tabela.to_html()))

C:\Users\Jhonny\AppData\Local\Temp\ipykernel_19116\979754042.py:49: FutureWarning: Styler.applymap has been deprecated. Use Styler.map instead.
  .applymap(colorir_valores, subset=['2017-2018', '2018-2019', '2019-2022'])


## 📊 Análise Final – Variação Percentual de Empréstimos (Pós-graduação)

O objetivo desta análise foi identificar a **variação percentual no volume de empréstimos realizados pelos alunos de pós-graduação**, nos cursos disponíveis, ao longo dos anos de **2017, 2018, 2019** e a **projeção de 2022**.

Os dados apresentados consideram:

- Apenas alunos de **pós-graduação**.
- Registros de empréstimos entre **2017 a 2019** (dados reais).
- Previsão para o ano de **2022**.

### 🎯 Principais objetivos:

- Fornecer insights à diretoria da biblioteca para **estratégias de marketing segmentadas**.
- Identificar cursos com **queda significativa** ou **crescimento consistente** de uso dos acervos.
- Direcionar ações de incentivo ou reforço no atendimento por curso.

### 📈 O que observar na tabela:

- As colunas representam a **variação percentual** entre pares de anos consecutivos.
- Valores em **verde** indicam aumento de empréstimos.
- Valores em **vermelho** indicam queda.
- As variações já estão formatadas com o símbolo “%” e arredondadas.

Essa tabela será incorporada ao sistema da equipe de Front-End, garantindo **padronização visual e clareza para os gestores** da biblioteca.

## 📊 Análise Final: Variação Percentual de Empréstimos (Pós-Graduação)

Com o objetivo de auxiliar a diretoria da biblioteca no planejamento de ações de marketing voltadas para os cursos de pós-graduação, realizamos uma análise comparativa das variações percentuais no número de empréstimos realizados entre os anos:

- 2017 → 2018
- 2018 → 2019
- 2019 → 2022 (previsão)

Os dados foram apresentados em uma tabela com valores percentuais, destacando em verde os aumentos e em vermelho as quedas nos empréstimos, conforme solicitado pela equipe de Front-end.

### 🔍 Principais Insights

- A maioria dos cursos apresentou queda consistente no número de empréstimos entre os anos analisados.
- O período de 2019 para 2022 mostra os impactos mais acentuados, com variações negativas acima de 70% em cursos como:
  - Administração (-74.98%)
  - Arquitetura e Urbanismo (-76.63%)
  - Engenharia Civil (-78.23%)
  - Filosofia (-75.38%)

- Algumas exceções positivas chamaram a atenção:
  - Bioinformática e Ciências Odontológicas apresentaram crescimento entre 2019 e 2022, com destaque para Bioinformática (+21.86%).
  - Gestão da Informação e do Conhecimento também teve ligeiro crescimento (+1.48%) em 2022.

### 💡 Possíveis Interpretações

- A queda brusca de 2019 para 2022 pode estar fortemente associada aos efeitos da pandemia da COVID-19, que alterou a dinâmica de ensino, restringiu o acesso físico à biblioteca e incentivou o uso de materiais digitais.
- Cursos que apresentaram resiliência ou crescimento podem estar mais conectados a áreas com forte base tecnológica ou acesso remoto a conteúdo digital.

### 🎯 Recomendações para a Diretoria

- Reforçar ações de incentivo à retomada do uso do acervo físico, especialmente nos cursos que sofreram maiores quedas.
- Investir em marketing direcionado para os cursos que mostraram resiliência, utilizando-os como referência de boas práticas.
- Investigar internamente os fatores que levaram ao crescimento de alguns cursos (ex: Bioinformática) e avaliar a replicabilidade dessas estratégias.
- Considerar parcerias com docentes e coordenações de curso para promover formações sobre o uso estratégico da biblioteca.